src/runners/rfq_maker_balance_cache_runner.ts

```diff
@@ -0,0 +1,169 @@
+): Promise<void> {
+    const workerId = _.uniqueId('rfqw_');
+    let lastBlockSeen = -1;
+    while (true) {
+        const newBlock = await ethersProvider.getBlockNumber();
+        if (lastBlockSeen < newBlock) {
+            lastBlockSeen = newBlock;
+            LATEST_BLOCK_PROCESSED_GAUGE.labels(workerId).set(lastBlockSeen);
+            logUtils.log(
+                {
+                    block: lastBlockSeen,
+                    workerId,
+                },
+                'Found new block',
+            );
+
+            const makerTokens = await _getMakerTokensAsync(connection);
+            const balancesCallInput = _splitValues(makerTokens);
+
+            const updateTime = new Date();
+            const erc20Balances = await _getErc20BalancesAsync(balanceCheckerContractInterface, balancesCallInput);
+
+            await _updateErc20BalancesAsync(balancesCallInput, erc20Balances, connection, updateTime);
+
+            await delay(DELAY_WHEN_NEW_BLOCK_FOUND);
+        } else {
+            await delay(DELAY_WHEN_NEW_BLOCK_NOT_FOUND);
+        }
+    }
+}
+
+// NOTE: this only returns a partial entity class, just token address and maker address
+// Cache the query results to reduce reads from the DB
+let MAKER_TOKEN_CACHE: ResultCache<MakerBalanceChainCacheEntity[]>;
+const _getMakerTokensAsync = async (connection: Connection) => {
+    if (!MAKER_TOKEN_CACHE) {
+        logUtils.log('updating cache');
+        MAKER_TOKEN_CACHE = createResultCache<any[]>(
+            () =>
+                connection
+                    .getRepository(MakerBalanceChainCacheEntity)
+                    .createQueryBuilder('maker_balance_chain_cache')
+                    .select(['maker_balance_chain_cache.tokenAddress', 'maker_balance_chain_cache.makerAddress'])
+                    .getMany(),
+            TEN_MINUTES_MS,
+        );
+    }
+    return (await MAKER_TOKEN_CACHE.getResultAsync()).result;
+};
+
+function _splitValues(makerTokens: MakerBalanceChainCacheEntity[]): BalancesCallInput {
+    const functionInputs: BalancesCallInput = { addresses: [], tokens: [] };
+
+    return makerTokens.reduce(({ addresses, tokens }, makerToken) => {
+        return {
+            addresses: addresses.concat(makerToken.makerAddress as string),
+            tokens: tokens.concat(makerToken.tokenAddress as string),
+        };
+    }, functionInputs);
+}
+
+async function _getBalanceCheckerContractInterfaceAsync(
+    provider: ethers.providers.JsonRpcProvider,
+): Promise<ethers.Contract> {
+    return new ethers.Contract(BALANCE_CHECKER_ADDRESS, BALANCE_CHECKER_ABI, provider);
+}
+
+async function _getErc20BalancesAsync(
+    balanceCheckerContractInterface: ethers.Contract,
+    balancesCallInput: BalancesCallInput,
+): Promise<string[]> {
+    // due to gas contraints limit the call to 1K balance checks
+    const addressesChunkedArray = _.chunk(balancesCallInput.addresses, MAX_BALANCE_CHECKS_PER_CALL);
```

Comment:
I appreciate that you are being mindful about the payload size. I agree this is something we need to be mindful about, in fact the current implementation could only really work if we have hundreds of addresses and not thousands. I would really encourage us to add a log warning somewhere if the number of addresses end up growing more than that number, since we may want to think about another way of doing this.

Reply:
Discussed, keeping this in for protection against high gas usage